### ISSUE TITLE
DL4006 should understand inheritance within a Dockerfile

### ISSUE THREAD

**Issue by docwhat** [hadolint/hadolint#406]
- [x] This is a bug report
- [ ] This is a feature request
- [ ] I searched existing issues before opening this one

### Expected behavior

If I set `SHELL ["/bin/bash", "-o", "pipefail"]` in a base target then I expect DL4006 not to trigger on any target building on it.

### Actual behavior

DL4006 triggers on the target building on the base target that has the `SHELL` line in it.

### Steps to reproduce the behavior

Run `hadolint` on the included Dockerfile.

**Output of `hadolint --version`:**

```bash
Haskell Dockerfile Linter 1.17.4-no-git
```

**Dockerfile** (if relevant)

```Dockerfile
FROM alpine:3 AS lower
SHELL ["/bin/sh", "-o", "pipefail"]

FROM lower AS upper
RUN true | cat
```

**Comment by eero-t** [comment]
This is not the only thing where hadolint fails to consider build stage inheritance.  It gives false warnings also for USER and WORKDIR.
